Firebase.h

```diff
@@ -25,49 +25,144 @@
+class FirebaseGetResult;
+class FirebasePushResult;
+class FirebaseRemoveResult;
+class FirebaseEventStream;
+
+// Primary client to the Firebase backend.
+class Firebase {
+ public:
+  Firebase(const String& host);
+  Firebase& auth(const String& auth);
+
+  // Fetch result at "path".
+  FirebaseGetResult get(const String& path);
+
+  // Add new value to list at "path", will return child name of new item.
+  FirebasePushResult push(const String& path, const String& value);
+
+  // Deletes value at "path" from server.
+  FirebaseRemoveResult remove(const String& path);
+
+  // Starts a stream of events that effect object at "path".
+  FirebaseEventStream stream(const String& path);
+
+ private:
+
+  int sendRequest(const char* method, const String& path, const String& value);
+
+  HTTPClient http_;
+  String host_;
+  String auth_;
+};
+
+// Base class for Results from a Firebase call.
+class FirebaseResult {
  public:
-  operator bool() const { return _code < 0; }
-  int code() const { return _code; }
-  const String& message() const { return _message; }
-  void reset() { set(0, ""); }
-  void set(int code, const String& message) {
-    _code = code;
-    _message = message;
+  // Constructor for error result.
+  FirebaseResult(const String& error_message);
+  // Constructor if no error.
+  FirebaseResult();
+
+  // True if no error.
+  operator bool() const;
+
+  bool isError() const;
+
+  const String& errorMessage() const;
+
+ private:
+  bool is_error_ = false;
+  String error_message_;
+};
+
+class FirebaseRemoveResult : public FirebaseResult {
+ public:
+  static FirebaseRemoveResult FromError(const String& error_message) {
+    return FirebaseRemoveResult(error_message);
+  }
+
+  static FirebaseRemoveResult Ok() {
+    return FirebaseRemoveResult();
   }
+
  private:
-  int _code = 0;
-  String _message = "";
+  FirebaseRemoveResult(const String& error_message);
+  FirebaseRemoveResult();
 };
 
-// Firebase is the connection to firebase.
-class Firebase {
+class FirebasePushResult : public FirebaseResult {
  public:
-  Firebase(const String& host);
-  Firebase& auth(const String& auth);
-  const FirebaseError& error() const {
-    return _error;
+  static FirebasePushResult FromError(const String& error_message) {
+    return FirebasePushResult(error_message);
   }
-  String get(const String& path);
-  String push(const String& path, const String& value);
-  bool connected();
-  Firebase& stream(const String& path);
-  bool available();
+
+  static FirebasePushResult FromResponse(const String& response) {
+    FirebasePushResult result;
+    // TODO(edcoyne): add json parsing to get name object.
+    result.name_ = response;
+    return result;
+  }
+
+  const String& name() const;
+
+ private:
+  FirebasePushResult(const String& error_message);
+  FirebasePushResult();
+
+  String name_;
+};
+
+class FirebaseGetResult : public FirebaseResult {
+ public:
+  static FirebaseGetResult FromError(const String& error_message) {
```

Comment:
We could make them private and make them friend classes. This way allows clients to create tests easily though. I don't feel strongly either way.